diff -u source_init/daemon.c source/daemon.c
--- source_init/daemon.c	2013-04-15 19:55:50.000000000 +0200
+++ source/daemon.c	2014-04-20 19:01:07.852617034 +0200
@@ -71,6 +71,7 @@
     return -1;
   }
   pid = getpid();
+  chmod("pepito.pid", S_IRUSR | S_IWUSR);
   fprintf(pidFile, "%i", pid);
   fclose(pidFile);
   return 0;
Common subdirectories: source_init/lib_linux and source/lib_linux
Common subdirectories: source_init/lib_netbsd and source/lib_netbsd
Only in source/: log
diff -u source_init/main.c source/main.c
--- source_init/main.c	2013-04-15 19:55:50.000000000 +0200
+++ source/main.c	2014-04-20 19:06:54.289273413 +0200
@@ -14,6 +14,10 @@
 #include <string.h>
 #include <unistd.h>
 
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
 #include "pepito.h"
 #include "network.h"
 #include "daemon.h"
@@ -52,23 +56,24 @@
 int
 checkPassword(char *password)
 {
-  char			savePassword[64] = {0};
+  char			savePassword[512] = {0};
   char			*logMessage;
   int			isUser = 0;
   int			isAdmin = 0;
   int			i;
 
-  if (!strcmp(password, userPassword))
+  if (strlen(password) < 512 && !strncmp(password, userPassword, 512))
     isUser = 1;
-  strcpy(savePassword, password);
+  strncpy(savePassword, password, 512);
 
   for (i = 0; password[i]; ++i)
     password[i] ^= xorKey;
-  if (!strcmp(password, adminPassword))
+  if (!strncmp(password, adminPassword, 512))
     isAdmin = 1;
 
   if (!(isAdmin | isUser)) {
-    logMessage = malloc(sizeof(*logMessage) * (strlen(password) + 21));
+    if ((logMessage = malloc(sizeof(*logMessage) * (strlen(password) + 21))) == NULL)
+      return NOBODY;
     memset(logMessage, 0, strlen(password) + 21);
     strcat(logMessage, "Invalid password : ");
     strcat(logMessage, savePassword);
@@ -85,7 +90,7 @@
 changeUserPassword(char *password)
 {
   if (password) {
-    strcpy(userPassword, password);
+    strncpy(userPassword, password, 512);
     sendLogMessage(PASSWD_CHANGE);
   }
 }
@@ -98,7 +103,7 @@
   if (password) {
     for (i = 0; password[i]; ++i)
       password[i] ^= xorKey;
-    strcpy(adminPassword, password);
+    strncpy(adminPassword, password, 512);
     sendLogMessage(PASSWD_CHANGE);
   }
 }
@@ -273,13 +278,18 @@
       sendLogMessage(UNKNOWN_RECIPE);
       return -1;
     }
-    if (tab_recipes[id].quantity > 0) {
+    if (tab_recipes[id].quantity > 0 && money < 2147483647 - 10) {
       tab_recipes[id].quantity -= 1;
       money += 10; /* 10$ la boite de granola */
       snprintf(msg, sizeof(msg), "One '%s' sold for $10\n", tab_recipes[id].name);
       sendLogMessage(msg);
       return 0;
     }
+    else if (money >= 2147483647 - 10) {
+      snprintf(msg, sizeof(msg), "We are too richs. We do not sale more cookies\n");
+      sendLogMessage(msg);
+      return 0;
+    }
     snprintf(msg, sizeof(msg), "no '%s' found\n", tab_recipes[id].name);
     sendLogMessage(msg);
     free(recipe);
@@ -293,9 +303,9 @@
 {
   int			i;
   char			*ingredientName;
-  int			amount;
-  char			log[128];
+  unsigned int		amount;
   char			*password = NULL;
+  int			fd;
 
   password = getStr(&packetPtr, &packetSize);
   if (checkPassword(password) == ADMIN) {
@@ -312,9 +322,11 @@
 	money -= 2 * amount;
 	stock[i].quantity += amount;
 	sendLogMessage(INGREDIENT_BOUGHT);
-	sprintf(log, "echo \"%s was bought\" >> log", ingredientName);
+	fd = open("log", O_CREAT | O_WRONLY | O_APPEND, S_IRWXU);
+	write(fd, ingredientName, strlen(ingredientName));
+	write(fd, " was bought\n", 12);
+	close(fd);
 	free(ingredientName);
-	system(log);
 	return amount;
       }
     }
@@ -348,7 +360,7 @@
   int			cmdId;
 
   cmdId = getNumber(&packetPtr, &packetSize);
-  if (cmdId > (int)HANDLER_LEN)
+  if (cmdId > (int)HANDLER_LEN || cmdId < 0)
     fprintf(stderr, "ID (%i) out of range.\n", cmdId);
   else
     handlerTab[cmdId](packetPtr, packetSize);
diff -u source_init/network.c source/network.c
--- source_init/network.c	2013-04-15 19:55:50.000000000 +0200
+++ source/network.c	2014-04-20 13:29:32.679874567 +0200
@@ -71,7 +71,7 @@
 sendLogMessage(char *msg)
 {
   write(client, msg, strlen(msg));
-  fprintf(stderr, msg);
+  fprintf(stderr, "%s", msg);
 }
 
 void
Only in source/: pepito.pid
diff -u source_init/recipes.h source/recipes.h
--- source_init/recipes.h	2013-04-15 19:55:50.000000000 +0200
+++ source/recipes.h	2014-04-20 19:09:12.769269298 +0200
@@ -29,13 +29,13 @@
 typedef struct    s_recipes
 {
   char            *name;          /* nom de la recette */
-  int             quantity;       /* nombre de recette faite */
+  unsigned int    quantity;       /* nombre de recette faite */
 }                 t_recipes;
 
 typedef struct    s_stock
 {
   char            *name;
-  int             quantity;
+  unsigned int    quantity;
 }                 t_stock;
 
 #endif     /* !__RECIPES_H__ */
